primitives/weights/src/weight_v2.rs

```diff
@@ -15,30 +15,40 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use codec::{CompactAs, Decode, Encode, MaxEncodedLen};
+use codec::{Compact, Decode, Encode, MaxEncodedLen};
 use core::ops::{Add, AddAssign, Div, Mul, Sub, SubAssign};
 use sp_arithmetic::traits::{Bounded, CheckedAdd, CheckedSub, Zero};
 use sp_debug_derive::RuntimeDebug;
 
 use super::*;
 
-#[derive(
-	Encode,
-	Decode,
-	MaxEncodedLen,
-	TypeInfo,
-	Eq,
-	PartialEq,
-	Copy,
-	Clone,
-	RuntimeDebug,
-	Default,
-	CompactAs,
-)]
+#[derive(Encode, MaxEncodedLen, TypeInfo, Eq, PartialEq, Copy, Clone, RuntimeDebug, Default)]
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 pub struct Weight {
+	#[codec(compact)]
 	/// The weight of computational time used based on some reference hardware.
 	ref_time: u64,
+	#[codec(compact)]
+	/// The weight of storage space used by proof of validity.
+	proof_size: u64,
+}
+
+// Custom Decode implementation for the purposes of migrating from 1D weights.
+impl Decode for Weight {
+	fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
+		let compact_ref_time = match Compact::<u64>::decode(input) {
+			Ok(r) => r,
+			Err(_) => {
+				// Try decoding it in the old way where we only had a non-compact ref time weight
+				let ref_time = u64::decode(input)
+					.map_err(|e| e.chain("Could not decode `Weight::ref_time`"))?;
+				return Ok(Self { ref_time, proof_size: 0 })
```

Comment:
You could write a small test that encoded a WeightV1 and tries to decode it as new WeightV2. Probably need to account for the compact?

Reply:
It is a bit difficult, since it requires a place where I can annotate `#[pallet::compact]`, and the only place I know is within the parameters of a dispatchable function. I'm not sure exactly how I can encode a `Weight` in the old way and try decoding it in the new way -- you can't really suddenly change the `Weight` parameter to a 2D weight type from a 1D one, yeah?

Reply:
I am not sure if i really understood your pain-point, but is this not enough?  
```rust
#[test]
fn compact_weight_v1_can_be_decoded_as_v2() {
	use codec::{Decode, Encode, Compact};
	type WeightV1 = u64;

	let weight_v1: WeightV1 = 12345;
	let compact_weight_v1 = Compact(weight_v1);
	let encoded = Compact::<WeightV1>::encode(&compact_weight_v1);

	// Decode as weight v2
	let decoded: Weight = Decode::decode(&mut &encoded[..]).unwrap();
	assert_eq!(decoded.ref_time(), weight_v1);
}
```

Reply:
It is not a bad unit test on its own, but it doesn't really test the `#[pallet::compact] OldWeight` to the `NewWeight` transition directly.